### ISSUE TITLE
Unify PropSchema and ModelSchema

### ISSUE THREAD

**Issue by NaridaL** [mobxjs/serializr#120]
There's currently a weird dichotomy with ModelSchema being a "serialization descriptor" for objects and PropSchema being one for all other types of values. There's `object()` which converts ModelSchema -> PropSchema, and the logic which would be in ModelSchema.serializer is instead the separate function "serialize".

This would be a breaking change, but as I've already got one in after the TypeScript conversion I'm considering making this change too. 

Basically it would change ModelSchema to `implements PropSchema`, deserialize would just create a new Context and then delegate to the passed "PropSchema". object() and createSimple/ModelSchema would be unified and contain the logic previously part of serialize and deserialize.

Any thoughts? @mweststrate @alexggordon

**Comment by mweststrate** [comment]
@NaridaL sounds good in the abstract, but I haven't been using or working on this package for a couple of years :) Would you mind giving a quick example of what the breaking would look like from consumer perspective?

**Comment by peey** [comment]
A quick note that this feature request (https://github.com/mobxjs/serializr/issues/147) might be an easy byproduct of the proposed unification.

So in case you move ahead with the unification, please keep this in mind.

**Comment by peey** [comment]
Also this unification would enable thing like `deserialize(list(A), [...stuff...])` or `deserialize(reference(User, findUserById), {...some serialize representation of id...})`

i.e. it'll enable re-use of already-declared logic even where you have contents of *just the property* with you, instead of a class that houses the property.

This re-use is incredibly useful when declaring `custom` serializers that want to invoke serialization/deserialization on a sub-part of the data.
